client/client.go

```diff
@@ -1,19 +1,307 @@
+	"go.mozilla.org/pkcs7"
 	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/protobuf/proto"
 )
 
-type Client struct {
-	bootz.BootstrapClient
+// Represents a 128 bit nonce.
+const nonceLength = 16
+
+var (
+	bootMode = flag.String("boot_mode", "SecureOnly", "The BootMode the device can start in.")
+	port     = flag.String("port", "", "The port to listen to on localhost for the bootz server.")
+	rootCA   = flag.String("root_ca_cert_path", "ca.pem", "The path to a file contained a PEM encoded certificate for the manufacturer CA.")
+)
+
+type OwnershipVoucher struct {
+	OV OwnershipVoucherInner `json:"ietf-voucher:voucher"`
+}
+
+// Defines the Ownership Voucher format. See https://www.rfc-editor.org/rfc/rfc8366.html.
+type OwnershipVoucherInner struct {
+	CreatedOn                  string `json:"created-on"`
+	ExpiresOn                  string `json:"expires-on"`
+	SerialNumber               string `json:"serial-number"`
+	Assertion                  string `json:"assertion"`
+	PinnedDomainCert           string `json:"pinned-domain-cert"`
+	DomainCertRevocationChecks bool   `json:"domain-cert-revocation-checks"`
+}
+
+// validateArtifacts checks the signed artifacts in a GetBootstrapDataResponse. Specifically, it:
+// - Checks that the OV in the response is signed by the manufacturer
+// - Checks that the serial number in the OV matches the one in the original request.
+// - Verififies that the Ownership Certificate is in the chain of signers of the Pinned Domain Cert.
+func validateArtifacts(serialNumber string, resp *bootz.GetBootstrapDataResponse, rootCA []byte) error {
+	ov64 := resp.GetOwnershipVoucher()
+	if len(ov64) == 0 {
+		return fmt.Errorf("received empty ownership voucher from server")
+	}
+
+	oc := resp.GetOwnershipCertificate()
+	if len(oc) == 0 {
+		return fmt.Errorf("received empty ownership certificate from server")
+	}
+	// Decode the ownership voucher
+	ov, err := base64.StdEncoding.DecodeString(string(ov64))
+	if err != nil {
+		return err
+	}
+
+	// Parse the PKCS7 message
+	p7, err := pkcs7.Parse(ov)
+	if err != nil {
+		return err
+	}
+	// Unmarshal the ownership voucher into a struct.
+	parsedOV := OwnershipVoucher{}
+	err = json.Unmarshal(p7.Content, &parsedOV)
+	if err != nil {
+		return err
+	}
+
+	// Create a CA pool for the device to validate that the vendor has signed this OV.
+	vendorCAPool := x509.NewCertPool()
+	if !vendorCAPool.AppendCertsFromPEM(rootCA) {
+		return fmt.Errorf("error adding vendor root CA to pool")
+	}
+
+	// Verify the ownership voucher with this CA.
+	err = p7.VerifyWithChain(vendorCAPool)
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Validated ownership voucher signed by vendor")
+
+	// Verify the serial number for this OV
+	if parsedOV.OV.SerialNumber != serialNumber {
+		return fmt.Errorf("serial number from OV does not match request")
+	}
+
+	// Get the Pinned Domain Cert from the Ownership Voucher.
```

Comment:
maybe move this to a function just ease of reading

Reply:
Is this just for the construction of the PDC or the rest of this function too?

Reply:
just the pdc piece

Reply:
Done, I think. Let me know if more should be pulled out from this function.